File: py/capacity.py
import sys
from typing import Iterable, List, Sequence, TextIO, Tuple, TypeVar

from utils import process_variant, mean, stdev, split
from vcf.population import Population, parse_population
from vcf.predictor import PredictorFactory, PredictorJoaoFactory
from vcf.stream import VCFStream


def compute_capacity(stream: VCFStream, population: Population, repeats: int, ratio: float, predictor_factory: PredictorFactory):
    individuals = stream.individuals
    split_index = int(len(individuals) * ratio)
    
    labels = population.get_labels(individuals)
    
    # Split once, since the call to `split` takes a lot of time to do for every
    # polymorphism and repeat. We call for the number of repeats and apply the
    # corresponding one, copying the same approach for all polymorphisms.
    if repeats > 1:
        train_target_indices_list = [
            split(range(len(individuals)), split_index)
            for _ in range(repeats)
        ]
        train_target_indices = [
            (set(i), set(j))
            for i, j in train_target_indices_list
        ]
    else:
        train_target_indices = [
            (set(range(i)), set(range(i)))
            for i, _ in enumerate(individuals)
        ]
    
    for polymorphism in stream:
        accuracies: List[float] = []
        
        for repeat_idx in range(repeats):
            genotypes = (process_variant(i) for i in polymorphism.data_fields)
            to_split = zip(individuals, genotypes, labels)
            
            train = []
            target = []
            for i, val in enumerate(to_split):
                if i in train_target_indices[repeat_idx][0]:
                    train.append(val)
                if i in train_target_indices[repeat_idx][1]:
                    target.append(val)
            
            _train_individuals, train_genotypes, train_labels = zip(*train)
            
            predictor = predictor_factory.build(train_genotypes, train_labels)
            
            count = 0.0
            for _target_individual, genotype, label in target:
                predictions = predictor.predict(genotype)
                if label in predictions:
                    count += predictions[label]
            
            accuracies.append(count / len(target))
        
        m = mean(accuracies)
        if repeats == 1:
            print(f'{polymorphism.get_identifier()}\t{m:.5f}')
        else:
            s = stdev(accuracies)
            print(f'{polymorphism.get_identifier()}\t{m:.5f}\t{s:.5f}')
        

if __name__ == '__main__':
    predictor_factory = PredictorJoaoFactory(0)
    
    with open('populations/superpopulations.txt') as f:
        population = parse_population(f)
    
    stream = VCFStream.from_text_stream(sys.stdin)
    compute_capacity(stream, population, 20, 0.1, predictor_factory)

File: rs/src/capacity.rs
use vcf::population::Population;
use vcf::predictor::{Predictor, PredictorFactory, PredictorJoaoFactory};
use vcf::stream::MetadataReader;

use utils::{process_variant, split, mean, stdev};

// use itertools::Itertools;
use std::collections::HashSet;

fn compute_capacity<S, PF>(
    mut stream: S,
    population: &Population,
    repeats: usize,
    ratio: f32,
    factory: PF,
) where
    S: MetadataReader,
    PF: PredictorFactory,
    <PF as PredictorFactory>::Predictor: Predictor<Prediction = f32>,
{
    let unfolded = stream.unfold().unwrap();

    let individuals = unfolded.individuals;
    let labels = population
        .labels(individuals.iter())
        .map(|s| s.to_string())
        .collect::<Vec<_>>();

    let split_index = (individuals.len() as f32 * ratio) as usize;

    let to_split = 0..individuals.len();
    let (train_idxs, target_idxs) = split(to_split, split_index);

    // Split once, since the call to `split` takes a lot of time to execute for
    // every polymorphism and every repeat. We call for the number of repeats
    // and apply the corresponding one, copying the same approach for all
    // polymorphisms.
    let train_target_indices: Vec<(HashSet<usize>, HashSet<usize>)>;

    if repeats > 1 {
        let train_target_indices_list = vec![split(0..individuals.len(), split_index); repeats];
        train_target_indices = train_target_indices_list
            .iter()
            .map(|(i, j)| (i.iter().cloned().collect(), j.iter().cloned().collect()))
            .collect();
    } else {
        train_target_indices = individuals
            .iter()
            .enumerate()
            .map(|(i, _)| ((0..i).collect(), (0..i).collect()))
            .collect();
    }
    
    let mut accuracies = Vec::with_capacity(repeats);
    
    for polymorphism in unfolded.stream {
        accuracies.clear();

        for repeat_idx in 0..repeats {
            let genotypes = polymorphism
                .genotypes()
                .map(|s| process_variant(s).unwrap_or(0.0))
                .collect::<Vec<_>>();
            let to_split = izip!(&individuals, genotypes, &labels);

            let mut train = Vec::new();
            let mut target = Vec::new();

            for (i, val) in to_split.enumerate() {
                if train_target_indices[repeat_idx].0.contains(&i) {
                    train.push(val);
                } else if train_target_indices[repeat_idx].1.contains(&i) {
                    target.push(val);
                }
            }

            let mut genotypes_and_labels = Vec::new();

            for (_individual, genotype, label) in train {
                genotypes_and_labels.push((genotype, label.to_string()));
            }

            let predictor = factory.build(genotypes_and_labels.iter().cloned());
            let mut count = 0.0;
            let total = target.len() as f32;

            for (_individual, genotype, label) in target {
                let predictions = predictor.predict(genotype);
                if let Some(prob) = predictions.get::<str>(&label) {
                    count += prob;
                }
            }

            accuracies.push(count / total);
        }

        let m = mean(&accuracies);
        if repeats == 1 {
            println!("{}\t{:.5}", polymorphism.identifier(), m);
        } else {
            let s = stdev(&accuracies);
            println!("{}\t{:.5}\t{:.5}", polymorphism.identifier(), m, s);
        }
    }
}

pub fn main() -> ::std::io::Result<()> {
    use std::fs::File;
    use std::io::{stdin, BufReader};
    use vcf::stream::from_text_stream;

    let predictor_factory = PredictorJoaoFactory { threshold: 0.0 };

    let population = {
        let file = File::open("../populations/superpopulations.txt")?;
        let reader = BufReader::new(file);
        Population::parse(reader)?
    };

    let stdin = stdin();
    let stream = {
        let handle = stdin.lock();
        from_text_stream(handle)
    };

    compute_capacity(stream, &population, 1, 0.1, predictor_factory);
    
    Ok(())
}
